pkg/cmd/membership/membership.go

```diff
@@ -0,0 +1,168 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/cmd/waitforkube"
+)
+
+const (
+	retryDuration     = 2 * time.Second
+	dialTimeout       = 5 * time.Second
+	etcdCertFile      = "/var/run/secrets/etcd-client/tls.crt"
+	etcdKeyFile       = "/var/run/secrets/etcd-client/tls.key"
+	etcdTrustedCAFile = "/var/run/configmaps/etcd-ca/ca-bundle.crt"
+	dbPath            = "member/snap/db"
+)
+
+type membership struct {
+	errOut         io.Writer
+	EtcdDnsName    string
+	EtcdEndpoints  string
+	etcdDataDir    string
+	EtcdDataFile   string
+	ExistingMember bool
+}
+
+// NewWaitForKubeCommand waits for kube to come up before continuing to start the rest of the containers.
+func NewMembershipCommand(errOut io.Writer) *cobra.Command {
+	membership := &membership{
+		errOut: errOut,
+	}
+	cmd := &cobra.Command{
+		Use:   "membership",
+		Short: "cluster membership validate",
+		Long:  "This command makes sure that the local etcd member is part of the cluster.",
+		Run: func(cmd *cobra.Command, args []string) {
+			must := func(fn func() error) {
+				if err := fn(); err != nil {
+					if cmd.HasParent() {
+						klog.Fatal(err)
+						fmt.Fprint(membership.errOut, err.Error())
+					}
+				}
+			}
+			must(membership.Complete)
+			must(membership.Validate)
+			must(membership.Run)
+		},
+	}
+	membership.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (m *membership) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&m.EtcdDnsName, "etcd-dns-name", "", "etcd member FQDN")
+	fs.StringVar(&m.EtcdEndpoints, "endpoints", "", "commas seperated list of etcd clients")
+	fs.StringVar(&m.etcdDataDir, "data-dir", "/var/lib/etcd", "path to etcd data-dir")
+}
+func (m *membership) Validate() error {
+	if m.IsExistingMember() {
+		if m.EtcdEndpoints == "" || m.EtcdDnsName == "" {
+			return fmt.Errorf("membership validation requires etcd-dns-name and endpoints")
+		}
+	}
+	return nil
+}
+
+func (m *membership) Complete() error {
+	m.EtcdDataFile = filepath.Join(m.etcdDataDir, dbPath)
+
+	etcdDnsName, etcdEndpoints := getENV()
+	// flags take priority over ENV if used.
+	if m.EtcdDnsName != "" && etcdDnsName != "" {
+		klog.Infof("flag --etcd-dns-name and ENV ETCD_DNS_NAME set, ENV will be ignored")
+	}
+	if m.EtcdEndpoints != "" && etcdEndpoints != "" {
+		klog.Infof("flag --endpoints and ENV ETCDCTL_ENDPOINTS set, ENV will be ignored")
+	}
+
+	if m.EtcdEndpoints == "" && etcdEndpoints != "" {
+		klog.Infof("recognized and used environment variable ETCDCTL_ENDPOINTS=%s", etcdEndpoints)
+		m.EtcdEndpoints = etcdEndpoints
+	}
+	if m.EtcdDnsName == "" && etcdDnsName != "" {
+		klog.Infof("recognized and used environment variable ETCD_DNS_NAME=%s", etcdDnsName)
+		m.EtcdDnsName = etcdDnsName
+	}
+	return nil
+}
+
+func (m *membership) Run() error {
+	if m.IsExistingMember() {
+		klog.Infof("etcd has previously been initialized as a member")
+		return nil
+	}
+
+	cli, err := getEtcdClient(strings.Split(m.EtcdEndpoints, ","))
```

Comment:
I think we should have a thin wrapper in our repo that deals with etcd client and use that for membership controller and here, we already have an issue tracking this, will update it with a reference

Reply:
yeah, in general, we have a lot of technical assets that can be centralized and resused.